Cache related issue with class definitions

Hi,

Since CGLIB 3.2.1+ (including 3.2.3) I experience the following error:

```
net.sf.cglib.core.CodeGenerationException: java.lang.reflect.InvocationTargetException-->null
	at net.sf.cglib.core.AbstractClassGenerator.generate(AbstractClassGenerator.java:332)
	at net.sf.cglib.proxy.Enhancer.generate(Enhancer.java:445)
	at net.sf.cglib.core.AbstractClassGenerator$ClassLoaderData$3.apply(AbstractClassGenerator.java:85)
	at net.sf.cglib.core.AbstractClassGenerator$ClassLoaderData$3.apply(AbstractClassGenerator.java:83)
	at net.sf.cglib.core.internal.LoadingCache$2.call(LoadingCache.java:54)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at net.sf.cglib.core.internal.LoadingCache.createEntry(LoadingCache.java:61)
	at net.sf.cglib.core.internal.LoadingCache.get(LoadingCache.java:34)
	at net.sf.cglib.core.AbstractClassGenerator$ClassLoaderData.get(AbstractClassGenerator.java:108)
	at net.sf.cglib.core.AbstractClassGenerator.create(AbstractClassGenerator.java:282)
	at net.sf.cglib.proxy.Enhancer.nextInstance(Enhancer.java:643)
	at net.sf.cglib.core.AbstractClassGenerator.create(AbstractClassGenerator.java:286)
	at net.sf.cglib.proxy.Enhancer.createHelper(Enhancer.java:433)
	at net.sf.cglib.proxy.Enhancer.create(Enhancer.java:304)
	at org.apache.wicket.jmx.Initializer.createProxy(Initializer.java:295)
	at org.apache.wicket.jmx.Initializer.register(Initializer.java:249)
	at org.apache.wicket.jmx.Initializer.init(Initializer.java:192)
	at org.apache.wicket.Application.initInitializers(Application.java:568)
	at org.apache.wicket.Application.initApplication(Application.java:776)
	at org.apache.wicket.protocol.http.WicketFilter.init(WicketFilter.java:431)
	at org.apache.wicket.protocol.http.WicketFilter.init(WicketFilter.java:355)
	at org.eclipse.jetty.servlet.FilterHolder.initialize(FilterHolder.java:137)
	at org.eclipse.jetty.servlet.ServletHandler.initialize(ServletHandler.java:856)
	at org.eclipse.jetty.servlet.ServletContextHandler.startContext(ServletContextHandler.java:346)
	at org.eclipse.jetty.webapp.WebAppContext.startWebapp(WebAppContext.java:1380)
	at org.eclipse.jetty.webapp.WebAppContext.startContext(WebAppContext.java:1342)
	at org.eclipse.jetty.server.handler.ContextHandler.doStart(ContextHandler.java:772)
	at org.eclipse.jetty.servlet.ServletContextHandler.doStart(ServletContextHandler.java:259)
	at org.eclipse.jetty.webapp.WebAppContext.doStart(WebAppContext.java:518)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:132)
	at org.eclipse.jetty.server.Server.start(Server.java:405)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:106)
	at org.eclipse.jetty.server.handler.AbstractHandler.doStart(AbstractHandler.java:61)
	at org.eclipse.jetty.server.Server.doStart(Server.java:372)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
	at org.apache.wicket.examples.StartExamples.main(StartExamples.java:110)
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at net.sf.cglib.core.ReflectUtils.defineClass(ReflectUtils.java:409)
	at net.sf.cglib.core.AbstractClassGenerator.generate(AbstractClassGenerator.java:323)
	... 36 more
Caused by: java.lang.LinkageError: loader (instance of  org/eclipse/jetty/webapp/WebAppClassLoader): attempted  duplicate class definition for name: "org/apache/wicket/jmx/Application"
	at java.lang.ClassLoader.defineClass1(Native Method)
	at java.lang.ClassLoader.defineClass(ClassLoader.java:763)
	... 42 more
```

There is no such error with 3.2.0.
The CGLIB related code that is involved is https://github.com/apache/wicket/blob/7973959f169a1c44c1de18a6acde21c8b926d64c/wicket-jmx/src/main/java/org/apache/wicket/jmx/Initializer.java#L256-L294

Any idea what could be the reason ?

--- a/cglib/src/main/java/net/sf/cglib/core/ReflectUtils.java
+++ b/cglib/src/main/java/net/sf/cglib/core/ReflectUtils.java
@@ -241,7 +241,9 @@ public class ReflectUtils {
             
         boolean flag = cstruct.isAccessible();
         try {
-            cstruct.setAccessible(true);
+            if (!flag) {
+                cstruct.setAccessible(true);
+            }
             Object result = cstruct.newInstance(args);
             return result;
         } catch (InstantiationException e) {
@@ -251,7 +253,9 @@ public class ReflectUtils {
         } catch (InvocationTargetException e) {
             throw new CodeGenerationException(e.getTargetException());
         } finally {
-            cstruct.setAccessible(flag);
+            if (!flag) {
+                cstruct.setAccessible(flag);
+            }
         }
                 
     }

--- a/cglib/src/main/java/net/sf/cglib/proxy/Enhancer.java
+++ b/cglib/src/main/java/net/sf/cglib/proxy/Enhancer.java
@@ -139,11 +139,6 @@ public class Enhancer extends AbstractClassGenerator
                                   Long serialVersionUID);
     }
 
-    /** Internal interface, only public due to ClassLoader issues. */
-    public interface EnhancerFactoryKey {
-        Object newInstance(EnhancerKey key);
-    }
-
     private Class[] interfaces;
     private CallbackFilter filter;
     private Callback[] callbacks;
@@ -392,6 +387,11 @@ public class Enhancer extends AbstractClassGenerator
         }
     }
 
+    /**
+     * The idea of the class is to cache relevant java.lang.reflect instances so
+     * proxy-class can be instantiated faster that when using {@link ReflectUtils#newInstance(Class, Class[], Object[])}
+     * and {@link Enhancer#setThreadCallbacks(Class, Callback[])}
+     */
     static class EnhancerFactoryData {
         public final Class generatedClass;
         private final Method setThreadCallbacks;
@@ -402,35 +402,43 @@ public class Enhancer extends AbstractClassGenerator
             this.generatedClass = generatedClass;
             try {
                 setThreadCallbacks = getCallbacksSetter(generatedClass, SET_THREAD_CALLBACKS_NAME);
-                setThreadCallbacks.setAccessible(true);
                 if (classOnly) {
                     this.primaryConstructorArgTypes = null;
                     this.primaryConstructor = null;
                 } else {
                     this.primaryConstructorArgTypes = primaryConstructorArgTypes;
-                    this.primaryConstructor = generatedClass.getDeclaredConstructor(primaryConstructorArgTypes);
-                    primaryConstructor.setAccessible(true);
+                    this.primaryConstructor = ReflectUtils.getConstructor(generatedClass, primaryConstructorArgTypes);
                 }
             } catch (NoSuchMethodException e) {
                 throw new CodeGenerationException(e);
             }
         }
 
+        /**
+         * Creates proxy instance for given argument types, and assigns the callbacks.
+         * Ideally, for each proxy class, just one set of argument types should be used,
+         * otherwise it would have to spend time on constructor lookup.
+         * Technically, it is a re-implementation of {@link Enhancer#createUsingReflection(Class)},
+         * with "cache {@link #setThreadCallbacks} and {@link #primaryConstructor}"
+         *
+         * @see #createUsingReflection(Class)
+         * @param argumentTypes constructor argument types
+         * @param arguments constructor arguments
+         * @param callbacks callbacks to set for the new instance
+         * @return newly created proxy
+         */
         public Object newInstance(Class[] argumentTypes, Object[] arguments, Callback[] callbacks) {
             setThreadCallbacks(callbacks);
             try {
-                if (!Arrays.equals(primaryConstructorArgTypes, argumentTypes)) {
-                    return ReflectUtils.newInstance(generatedClass, argumentTypes, arguments);
+                // Explicit reference equality is added here just in case Arrays.equals does not have one
+                if (primaryConstructorArgTypes == argumentTypes ||
+                        Arrays.equals(primaryConstructorArgTypes, argumentTypes)) {
+                    // If we have relevant Constructor instance at hand, just call it
+                    // This skips "get constructors" machinery
+                    return ReflectUtils.newInstance(primaryConstructor, arguments);
                 }
-                try {
-                    return primaryConstructor.newInstance(arguments);
-                } catch (InstantiationException e) {
-                    throw new CodeGenerationException(e);
-                } catch (IllegalAccessException e) {
-                    throw new CodeGenerationException(e);
-                } catch (InvocationTargetException e) {
-                    throw new CodeGenerationException(e.getTargetException());
-                }
+                // Take a slow path if observing unexpected argument types
+                return ReflectUtils.newInstance(generatedClass, argumentTypes, arguments);
             } finally {
                 // clear thread callbacks to allow them to be gc'd
                 setThreadCallbacks(null);
@@ -646,8 +654,17 @@ public class Enhancer extends AbstractClassGenerator
             throw new IllegalArgumentException("No visible constructors in " + sc);
     }
 
+    /**
+     * This method should not be called in regular flow.
+     * Technically speaking {@link #wrapCachedClass(Class)} uses {@link EnhancerFactoryData} as a cache value,
+     * and the latter enables faster instantiation than plain old reflection lookup and invoke.
+     * This method is left intact for backward compatibility reasons: just in case it was ever used.
+     *
+     * @param type class to instantiate
+     * @return newly created proxy instance
+     * @throws Exception if something goes wrong
+     */
     protected Object firstInstance(Class type) throws Exception {
-        // Should not reach here
         if (classOnly) {
             return type;
         } else {
@@ -656,10 +673,6 @@ public class Enhancer extends AbstractClassGenerator
     }
 
     protected Object nextInstance(Object instance) {
-        if (currentKey instanceof EnhancerFactoryKey) {
-            return instance;
-        }
-
         EnhancerFactoryData data = (EnhancerFactoryData) instance;
         if (classOnly) {
             return data.generatedClass;
@@ -676,9 +689,6 @@ public class Enhancer extends AbstractClassGenerator
 
     @Override
     protected Object wrapCachedClass(Class klass) {
-        if (currentKey instanceof EnhancerFactoryKey) {
-            return new WeakReference(ReflectUtils.newInstance(klass));
-        }
         Class[] argumentTypes = this.argumentTypes;
         if (argumentTypes == null) {
             argumentTypes = Constants.EMPTY_CLASS_ARRAY;
@@ -781,6 +791,15 @@ public class Enhancer extends AbstractClassGenerator
         return type.getDeclaredMethod(methodName, new Class[]{ Callback[].class });
     }
 
+    /**
+     * Instantiates a proxy instance and assigns callback values.
+     * Implementation detail: java.lang.reflect instances are not cached, so this method should not
+     * be used on a hot path.
+     * This method is used when {@link #setUseCache(boolean)} is set to {@code false}.
+     *
+     * @param type class to instantiate
+     * @return newly created instance
+     */
     private Object createUsingReflection(Class type) {
         setThreadCallbacks(type, callbacks);
         try{
@@ -797,14 +816,6 @@ public class Enhancer extends AbstractClassGenerator
         }finally{
          // clear thread callbacks to allow them to be gc'd
          setThreadCallbacks(type, null);
-        }
-    }
-
-    private Object createUsingFactory(Factory factory) {
-        if (argumentTypes != null) {
-            return factory.newInstance(argumentTypes, arguments, callbacks);
-        } else {
-            return factory.newInstance(callbacks);
         }
     }
 

--- a/cglib/src/main/java/net/sf/cglib/proxy/Enhancer.java
+++ b/cglib/src/main/java/net/sf/cglib/proxy/Enhancer.java
@@ -71,8 +71,6 @@ public class Enhancer extends AbstractClassGenerator
     private static final Source SOURCE = new Source(Enhancer.class.getName());
     private static final EnhancerKey KEY_FACTORY =
       (EnhancerKey)KeyFactory.create(EnhancerKey.class, KeyFactory.HASH_ASM_TYPE, null);
-    private static final EnhancerFactoryKey FACTORY_KEY_FACTORY =
-      (EnhancerFactoryKey)KeyFactory.create(EnhancerFactoryKey.class);
 
     private static final String BOUND_FIELD = "CGLIB$BOUND";
     private static final String FACTORY_DATA_FIELD = "CGLIB$FACTORY_DATA";
@@ -152,7 +150,6 @@ public class Enhancer extends AbstractClassGenerator
     private Type[] callbackTypes;
     private boolean validateCallbackTypes;
     private boolean classOnly;
-    private boolean factoryOnly;
     private Class superclass;
     private Class[] argumentTypes;
     private Object[] arguments;
@@ -299,7 +296,6 @@ public class Enhancer extends AbstractClassGenerator
      */
     public Object create() {
         classOnly = false;
-        factoryOnly = false;
         argumentTypes = null;
         return createHelper();
     }
@@ -315,7 +311,6 @@ public class Enhancer extends AbstractClassGenerator
      */
     public Object create(Class[] argumentTypes, Object[] arguments) {
         classOnly = false;
-        factoryOnly = false;
         if (argumentTypes == null || arguments == null || argumentTypes.length != arguments.length) {
             throw new IllegalArgumentException("Arguments must be non-null and of equal length");
         }
@@ -334,21 +329,7 @@ public class Enhancer extends AbstractClassGenerator
      */
     public Class createClass() {
         classOnly = true;
-        factoryOnly = false;
         return (Class)createHelper();
-    }
-
-    /**
-     * Generate a {@link Factory} implementation if necessary and return it.
-     * This ignores any callbacks that have been set.
-     * @see Factory#newInstance(Callback)
-     * @see Factory#newInstance(Callback[])
-     * @see Factory#newInstance(Class[], Object[], Callback[])
-     */
-    public Factory createFactory() {
-        classOnly = false;
-        factoryOnly = true;
-        return (Factory)createHelper();
     }
 
     /**
@@ -413,10 +394,58 @@ public class Enhancer extends AbstractClassGenerator
 
     static class EnhancerFactoryData {
         public final Class generatedClass;
-        public volatile Factory factory;
+        private final Method setThreadCallbacks;
+        private final Class[] primaryConstructorArgTypes;
+        private final Constructor primaryConstructor;
 
-        public EnhancerFactoryData(Class generatedClass) {
+        public EnhancerFactoryData(Class generatedClass, Class[] primaryConstructorArgTypes, boolean classOnly) {
             this.generatedClass = generatedClass;
+            try {
+                setThreadCallbacks = getCallbacksSetter(generatedClass, SET_THREAD_CALLBACKS_NAME);
+                setThreadCallbacks.setAccessible(true);
+                if (classOnly) {
+                    this.primaryConstructorArgTypes = null;
+                    this.primaryConstructor = null;
+                } else {
+                    this.primaryConstructorArgTypes = primaryConstructorArgTypes;
+                    this.primaryConstructor = generatedClass.getDeclaredConstructor(primaryConstructorArgTypes);
+                    primaryConstructor.setAccessible(true);
+                }
+            } catch (NoSuchMethodException e) {
+                throw new CodeGenerationException(e);
+            }
+        }
+
+        public Object newInstance(Class[] argumentTypes, Object[] arguments, Callback[] callbacks) {
+            setThreadCallbacks(callbacks);
+            try {
+                if (!Arrays.equals(primaryConstructorArgTypes, argumentTypes)) {
+                    return ReflectUtils.newInstance(generatedClass, argumentTypes, arguments);
+                }
+                try {
+                    return primaryConstructor.newInstance(arguments);
+                } catch (InstantiationException e) {
+                    throw new CodeGenerationException(e);
+                } catch (IllegalAccessException e) {
+                    throw new CodeGenerationException(e);
+                } catch (InvocationTargetException e) {
+                    throw new CodeGenerationException(e.getTargetException());
+                }
+            } finally {
+                // clear thread callbacks to allow them to be gc'd
+                setThreadCallbacks(null);
+            }
+
+        }
+
+        private void setThreadCallbacks(Callback[] callbacks) {
+            try {
+                setThreadCallbacks.invoke(generatedClass, (Object) callbacks);
+            } catch (IllegalAccessException e) {
+                throw new CodeGenerationException(e);
+            } catch (InvocationTargetException e) {
+                throw new CodeGenerationException(e.getTargetException());
+            }
         }
     }
 
@@ -635,19 +664,14 @@ public class Enhancer extends AbstractClassGenerator
         if (classOnly) {
             return data.generatedClass;
         }
-        Factory factory = data.factory;
-        if (factory == null) {
-            this.currentData = data;
-            Object factoryKey = FACTORY_KEY_FACTORY.newInstance((EnhancerKey) currentKey);
-            this.currentKey = factoryKey;
-            factory = (Factory) super.create(factoryKey);
-            this.currentData = null;
-            data.factory = factory;
+
+        Class[] argumentTypes = this.argumentTypes;
+        Object[] arguments = this.arguments;
+        if (argumentTypes == null) {
+            argumentTypes = Constants.EMPTY_CLASS_ARRAY;
+            arguments = null;
         }
-        if (factoryOnly) {
-            return factory;
-        }
-        return createUsingFactory(factory);
+        return data.newInstance(argumentTypes, arguments, callbacks);
     }
 
     @Override
@@ -655,7 +679,11 @@ public class Enhancer extends AbstractClassGenerator
         if (currentKey instanceof EnhancerFactoryKey) {
             return new WeakReference(ReflectUtils.newInstance(klass));
         }
-        EnhancerFactoryData factoryData = new EnhancerFactoryData(klass);
+        Class[] argumentTypes = this.argumentTypes;
+        if (argumentTypes == null) {
+            argumentTypes = Constants.EMPTY_CLASS_ARRAY;
+        }
+        EnhancerFactoryData factoryData = new EnhancerFactoryData(klass, argumentTypes, classOnly);
         Field factoryDataField = null;
         try {
             factoryDataField = klass.getField(FACTORY_DATA_FIELD);
@@ -798,7 +826,7 @@ public class Enhancer extends AbstractClassGenerator
      * Helper method to create an intercepted object.
      * For finer control over the generated instance, use a new instance of <code>Enhancer</code>
      * instead of this static method.
-     * @param type class to extend or interface to implement
+     * @param superclass class to extend or interface to implement
      * @param interfaces array of interfaces to implement, or null
      * @param callback the callback to use for all methods
      */
@@ -814,7 +842,7 @@ public class Enhancer extends AbstractClassGenerator
      * Helper method to create an intercepted object.
      * For finer control over the generated instance, use a new instance of <code>Enhancer</code>
      * instead of this static method.
-     * @param type class to extend or interface to implement
+     * @param superclass class to extend or interface to implement
      * @param interfaces array of interfaces to implement, or null
      * @param filter the callback filter to use when generating a new class
      * @param callbacks callback implementations to use for the enhanced object

--- a/cglib/src/test/java/net/sf/cglib/proxy/TestEnhancer.java
+++ b/cglib/src/test/java/net/sf/cglib/proxy/TestEnhancer.java
@@ -657,6 +657,26 @@ public class TestEnhancer extends CodeGenTestCase {
       assertEquals(name + "45", proxied2.getCanonicalName());
     }
 
+    /**
+     * In theory, every sane implementation of {@link NamingPolicy} should check if the class name is occupied,
+     * however, in practice there are implementations in the wild that just return whatever they feel is good.
+     *
+     * @throws Throwable if something wrong happens
+     */
+    public void testNamingPolicyThatReturnsConstantNames() throws Throwable {
+      Enhancer e = new Enhancer();
+      final String desiredClassName = "net.sf.cglib.empty.Object$$42";
+      e.setCallback(NoOp.INSTANCE);
+      e.setClassLoader(new ClassLoader(this.getClass().getClassLoader()){});
+      e.setNamingPolicy(new NamingPolicy() {
+        public String getClassName(String prefix, String source, Object key, Predicate names) {
+          return desiredClassName;
+        }
+      });
+      Class proxied = e.create().getClass();
+      assertEquals("Class name should match the one returned by NamingPolicy", desiredClassName, proxied.getName());
+    }
+
     public static Object enhance(Class cls, Class interfaces[], Callback callback, ClassLoader loader) {
         Enhancer e = new Enhancer();
         e.setSuperclass(cls);

--- a/cglib/src/test/java/net/sf/cglib/proxy/TestEnhancer.java
+++ b/cglib/src/test/java/net/sf/cglib/proxy/TestEnhancer.java
@@ -161,13 +161,6 @@ public class TestEnhancer extends CodeGenTestCase {
 
         Factory factory = (Factory)proxy;
         assertEquals("((EA)factory.newInstance(factory.getCallbacks())).getName()", "herby", ((EA)factory.newInstance(factory.getCallbacks())).getName());
-
-        Enhancer e = new Enhancer();
-        e.setSuperclass(EA.class);
-        e.setCallbackType(MethodInterceptor.class);
-        Factory factory1 = e.createFactory();
-        assertEquals("((EA)e.createFactory().newInstance(factory.getCallbacks())).getName()",
-                "herby", ((EA)factory1.newInstance(new DelegateInterceptor(save))).getName());
     }
 
     class DelegateInterceptor implements MethodInterceptor {

